src/CRISP.BackendChallenge/Controllers/EmployeeController.cs

```diff
@@ -7,33 +7,266 @@
+            return CreateInternalServerError();
+        }
+    }
+
+    
+    [HttpGet]
+    [Route("search")]
+    public IActionResult SearchEmployees(int? id, string? name, int? department)
+    {
+        _logger.LogDebug(":: Performing {MethodName}", nameof(SearchEmployees));
+
+        try
+        {
+            if (id == null && string.IsNullOrWhiteSpace(name) && department == null)
+            {
+                var employees = _employeeRepository.GetAll();
+
+                var res = employees.Select(ToEmployeeResponse);
+
+                return Ok(res);
+            }
+            
+            var result = _employeeRepository.Query().Include(e => e.Logins)
+                .Where(x => (id.HasValue && x.Id == id) || (!string.IsNullOrEmpty(name) && x.Name == name) ||
+                            (department.HasValue && x.Department == (Department)department))
+                .Select(x => ToEmployeeResponse(x)).ToList();
+            
+            return Ok(result);
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(":: Error occurred while searching employees: {EMessage}", e.Message);
+            return CreateInternalServerError();
+        }
+        
+    }
+    
+    
+    [HttpPut]
+    [Route("update/{id:int}")]
+    public IActionResult UpdatedById(int id,[FromBody] EmployeeRequest employeeRequest)
+    {
+        _logger.LogDebug(":: Performing {MethodName} with ID {Id}", nameof(UpdatedById), id);
+        
+        try
+        {
+           
+            var existingEmployee = _employeeRepository.GetById(id);
+            
+            existingEmployee.Name = employeeRequest.Name ?? existingEmployee.Name;
+            if (employeeRequest.Department != null)
             {
-                Id = x.Id,
-                Name = x.Name,
-                // TODO: Include the login date information...
-                LoginDates = null
-            });
-        return Ok(result);
-    }
-    // TODO: Implement Search By Id
-    // TODO: Implement Delete by Id
-    // TODO: Implement Update by Id
+                existingEmployee.Department = (Department)employeeRequest.Department;
+            }
+            existingEmployee.StartDate = employeeRequest.StartDate;
+            existingEmployee.EndDate = employeeRequest.EndDate;
+            
+            _employeeRepository.Update(existingEmployee);
+            _employeeRepository.Save();
+            
+            if (employeeRequest.LoginDates != null)
+            {
+                foreach (var login in employeeRequest.LoginDates.Select(l => new Login()
+                         {
+                             EmployeeId = existingEmployee.Id,
+                             LoginDate = l
+                         }))
+                {
+                    _loginRepository.Add(login);
+                }
+                _loginRepository.Save();
+            }
+
+            return Ok(ToEmployeeResponse(existingEmployee));
```

Comment:
You should create a new `Update` repository method to encapsulate the update logic. Rather than reusing any other repository methods (such as `GetById`), it should use its own queries for everything. Code duplication is ok here, because it allows you to fully segregate your logic between repository methods. The benefit of the repository pattern is to fully encapsulate the logic for each command or query

Reply:
Believe I fixed the Update method to make it more streamlined and understandable from the controller and repository perspectives